File: src/main/java/com/flaghacker/uttt/games/match/ConsoleGame.java
package com.flaghacker.uttt.games.match;

import com.flaghacker.uttt.common.Board;
import com.flaghacker.uttt.common.Bot;
import com.flaghacker.uttt.common.Coord;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

import static com.flaghacker.uttt.common.Board.ENEMY;
import static com.flaghacker.uttt.common.Board.PLAYER;

public class ConsoleGame
{
	private Bot bot;
	private Scanner scanner;

	private List<Board> history = new ArrayList<>();
	private Board curr = new Board();

	public ConsoleGame(Bot bot)
	{
		this.bot = bot;
		this.scanner = new Scanner(System.in);
	}

	public void run()
	{
		System.out.println(String.format("You (X) vs %s (O)", bot));
		System.out.println("\t<undo> to undo");
		System.out.println("\t<x>,<y> to play at those coordinates");

		while (true)
		{
			System.out.println("Current board: ");
			System.out.println(curr);
			String input = scanner.nextLine();

			if (input.equals("undo"))
			{
				if (history.size() == 0)
					System.err.println("no history");

				curr = history.get(history.size() - 1);
			}
			else if (input.matches("\\d,\\d"))
			{
				String[] arr = input.split(",");
				Coord coord = Coord.coord(Integer.parseInt(arr[0]), Integer.parseInt(arr[1]));

				if (curr.availableMoves().contains(coord))
				{
					System.out.println("move on " + coord);
					history.add(curr.copy());
					curr.play(coord, PLAYER);

					Coord botMove = bot.move(curr.copy());
					System.out.println("bot moves " + botMove);
					curr.play(botMove, ENEMY);
				}
				else
				{
					if (curr.isDone())
						System.out.println("game done, won by " + toSymbol(curr.wonBy()));
					else
						System.out.println("invalid move, choose one of " + curr.availableMoves());
				}
			}
			else
			{
				System.out.println("invalid command");
			}
		}
	}

	private static String toSymbol(byte player)
	{
		if (player == PLAYER)
			return "X";
		else if (player == ENEMY)
			return "O";
		else
			return "?";
	}
}

File: src/main/kotlin/games/ConsoleGame.kt
package com.flaghacker.sttt.games

import com.flaghacker.sttt.common.Board
import com.flaghacker.sttt.common.Bot
import com.flaghacker.sttt.common.Coord
import java.util.*

class ConsoleGame(private val bot: Bot) {
	private var current = Board()
	private val history = mutableListOf<Board>()

	fun run() {
		println("You (X) vs $bot (O)")
		printHelp()

		while (!current.isDone) {
			val input = Scanner(System.`in`).nextLine()

			when (input.split(" ").first().toLowerCase()) {
				"undo" -> {
					if (history.size > 1) {
						history.removeAt(history.lastIndex)
						current = history.last().copy()
						printCurrent()
					} else {
						history.clear()
						current = Board()
						printCurrent()
					}
				}
				"exit" -> return
				"help" -> printHelp()
				"play" -> {
					val coord = coordFromInput(input)
					if (coord != null) {
						if (current.availableMoves.contains(coord)) {
							println("you played on $coord")
							current.play(coord)
							printCurrent()

							if (!current.isDone) {
								val botMove = bot.move(current.copy())!!
								println("bot played on $botMove")
								current.play(botMove)
								history.add(current.copy())
								printCurrent()
							}
						} else System.err.println("coord not available please choose out of ${Arrays.toString(current.availableMoves)}")
					}
				}
				else -> {
					System.err.println("invalid command")
					printCurrent()
				}
			}
		}
	}

	private fun printCurrent() {
		println("\n${current.toString(true)}\n")
	}

	private fun printHelp() {
		println("\nCOMMANDS:")
		println("help\t\t\t: display the commands")
		println("undo\t\t\t: undo your last move")
		println("play <os>\t\t: play os [0-8] in current macro")
		println("play <om> <os>\t: play os [0-8] in macro om [0-8]")
		printCurrent()
	}

	private fun coordFromInput(input: String): Coord? {
		when {
			input.matches(Regex("play \\d")) -> {
				val os = input.split(" ").last().toInt()

				val macros = current.availableMoves.map { it / 9 }
				return if (macros.all { it == macros.first() })
					(macros.first() * 9 + os).toByte()
				else {
					System.err.println("you are in freemove, please use \"play <om> <os>\"")
					null
				}
			}
			input.matches(Regex("play \\d \\d")) -> {
				return (input.split(" ")[1].toInt() * 9 + input.split(" ")[2].toInt()).toByte()
			}
			else -> {
				System.err.println("play command formatted incorrectly, enter help for command list")
				return null
			}
		}
	}
}
